osu.Server.Queues.ScoreStatisticsProcessor/Processors/UserRankCountProcessor.cs

```diff
@@ -0,0 +1,118 @@
+                var bestScore = getBestScore(score.UserID, score.BeatmapID, score.RulesetID, excludedScoreId: null, conn, transaction);
+
+                if (bestScore?.ID != score.ID)
+                {
+                    // this score isn't the user's best on the beatmap, so nothing needs to be reverted.
+                    return;
+                }
+
+                updateRankCounts(userStats, score.Rank, revert: true);
+
+                // this score is the user's best, so fetch the next best (excluding it) so that we can apply the rank from that score.
+                var secondBestScore = getBestScore(score.UserID, score.BeatmapID, score.RulesetID, excludedScoreId: score.ID, conn, transaction);
+                if (secondBestScore != null)
+                    updateRankCounts(userStats, secondBestScore.Rank, revert: false);
+            }
+        }
+
+        public void ApplyToUserStats(SoloScoreInfo score, UserStats userStats, MySqlConnection conn, MySqlTransaction transaction)
+        {
+            // while looking for the best score, exclude the score being processed right now.
+            // this does nothing on the first process (as `getBestScore()` already only looks at scores processed at least once),
+            // but is important on subsequent reprocessings during `processed_version` bumps.
+            var bestScore = getBestScore(score.UserID, score.BeatmapID, score.RulesetID, excludedScoreId: score.ID, conn, transaction);
+
+            if (bestScore == null)
+            {
+                updateRankCounts(userStats, score.Rank, revert: false);
+                return;
+            }
+
+            if (score.TotalScore < bestScore.TotalScore) return;
+            if (score.TotalScore == bestScore.TotalScore && score.ID <= bestScore.ID) return;
+
+            updateRankCounts(userStats, bestScore.Rank, revert: true);
+            updateRankCounts(userStats, score.Rank, revert: false);
+        }
+
+        public void ApplyGlobal(SoloScoreInfo score, MySqlConnection conn)
+        {
+        }
+
+        private static SoloScoreInfo? getBestScore(int userId, int beatmapId, int rulesetId, ulong? excludedScoreId, MySqlConnection conn, MySqlTransaction transaction)
+        {
+            // purpose of join with `process_history` is to only look for already-processed scores
+            var rankSource = conn.QueryFirstOrDefault<SoloScore?>(
+                $"SELECT * FROM {SoloScore.TABLE_NAME} `s`"
+                + $"JOIN {ProcessHistory.TABLE_NAME} `ph` ON `ph`.`score_id` = `s`.`id`"
+                + "WHERE (@ExcludedScoreId IS NULL OR `s`.`id` <> @ExcludedScoreId) "
+                + "AND `s`.`user_id` = @UserId "
+                + "AND `s`.`beatmap_id` = @BeatmapId "
+                + "AND `s`.`ruleset_id` = @RulesetId "
+                + "AND `s`.`preserve` = 1 "
+                + "ORDER BY JSON_EXTRACT(`s`.`data`, '$.total_score') DESC, `s`.`id` DESC", new
```

Comment:
This `JSON_EXTRACT()` is probably not going to be a good time. Not sure what should happen (extract to column? some other mysql magic? on MSSQL I'd slap on a computed column and an index on that, but that's not a thing on mysql apparently)

Reply:
It's fine, since we're already filtering down to per-beatmap scores. Generally we should be working with well under 100 scores at this point.